File: docs/public/2025-12-21-leetcoding/dynamic-programming-explained/templates/c99/dp-15-word-break.c
/*
Title: Word Break
Problem: Given a string s and dictionary words, determine if s can be segmented into a sequence of dictionary words.
Constraints: 1 <= s.length <= 20, dictionary size <= 10, lowercase letters
Examples:
- Input: s="leetcode", dict=["leet","code"] => Output: true
- Input: s="catsandog", dict=["cats","dog","sand","and","cat"] => Output: false
- Input: s="cars", dict=["car","ca","rs"] => Output: true
Edge cases: Single character, overlapping words.

Test design coverage:
- Boundary tests: smallest sizes, zeros/empty when allowed, single-row/column when applicable.
- Equivalence classes: reachable vs unreachable states, varied distributions of values.
- Pitfall regression: known wrong iteration orders or invalid transitions.
- Randomized small tests with oracle.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int passed = 0;
static int failed = 0;

#define ASSERT_INT_EQ(actual, expected, msg) do { \
  if ((actual) != (expected)) { \
    failed += 1; \
    printf("FAIL: %s (expected %d, got %d)\\n", msg, (expected), (actual)); \
  } else { \
    passed += 1; \
  } \
} while (0)

#define ASSERT_BOOL_EQ(actual, expected, msg) ASSERT_INT_EQ((actual) != 0, (expected) != 0, msg)

static unsigned int seed = 123456789u;
static unsigned int lcg(void) {
  seed = seed * 1664525u + 1013904223u;
  return seed;
}

static int rand_int(int min, int max) {
  return min + (int)(lcg() % (unsigned int)(max - min + 1));
}


int solve(const char *s, const char **dict, int dictSize) {
  // TODO: implement
  return 0;
}


static int has_word(const char *s, int start, int end, const char **dict, int dictSize) {
  int len = end - start;
  for (int i = 0; i < dictSize; i += 1) {
    if ((int)strlen(dict[i]) != len) continue;
    if (strncmp(s + start, dict[i], len) == 0) return 1;
  }
  return 0;
}
static int ref_solve(const char *s, const char **dict, int dictSize) {
  int n = (int)strlen(s);
  int *dp = (int *)calloc(n + 1, sizeof(int));
  dp[0] = 1;
  for (int i = 1; i <= n; i += 1) {
    for (int j = 0; j < i; j += 1) {
      if (dp[j] && has_word(s, j, i, dict, dictSize)) {
        dp[i] = 1;
        break;
      }
    }
  }
  int ans = dp[n];
  free(dp);
  return ans;
}


static void test_deterministic(void) {

  {
    const char *dict[] = {"leet", "code"};
    ASSERT_BOOL_EQ(solve("leetcode", dict, 2), 1, "deterministic");
  }
  {
    const char *dict[] = {"apple", "pen"};
    ASSERT_BOOL_EQ(solve("applepenapple", dict, 2), 1, "deterministic");
  }
  {
    const char *dict[] = {"cats", "dog", "sand", "and", "cat"};
    ASSERT_BOOL_EQ(solve("catsandog", dict, 5), 0, "deterministic");
  }
  {
    const char *dict[] = {"car", "ca", "rs"};
    ASSERT_BOOL_EQ(solve("cars", dict, 3), 1, "deterministic");
  }
  {
    const char *dict[] = {"aaaa", "aaa"};
    ASSERT_BOOL_EQ(solve("aaaaaaa", dict, 2), 1, "deterministic");
  }
  {
    const char *dict[] = {"b"};
    ASSERT_BOOL_EQ(solve("a", dict, 1), 0, "deterministic");
  }

}

static void test_random(void) {

  const char letters[] = "abc";
for (int t = 0; t < 20; t += 1) {
  int dictSize = rand_int(1, 6);
  const char **dict = (const char **)malloc(sizeof(char *) * dictSize);
  char dictStorage[6][6];
  for (int i = 0; i < dictSize; i += 1) {
    int len = rand_int(1, 4);
    for (int k = 0; k < len; k += 1) dictStorage[i][k] = letters[rand_int(0, 2)];
    dictStorage[i][len] = '\0';
    dict[i] = dictStorage[i];
  }
  int sLen = rand_int(1, 8);
  char s[10];
  for (int k = 0; k < sLen; k += 1) s[k] = letters[rand_int(0, 2)];
  s[sLen] = '\0';
  int expected = ref_solve(s, dict, dictSize);
  ASSERT_BOOL_EQ(solve(s, dict, dictSize), expected, "random word break");
  free(dict);
}

}

int main(void) {

  test_deterministic();

  test_random();

  printf("Passed: %d, Failed: %d\n", passed, failed);

  return failed > 0 ? 1 : 0;

}

File: docs/public/2025-12-21-leetcoding/dynamic-programming-explained/templates/js/dp-15-word-break.js
/*
Title: Word Break
Problem: Given a string s and dictionary words, determine if s can be segmented into a sequence of dictionary words.
Constraints: 1 <= s.length <= 20, dictionary size <= 10, lowercase letters
Examples:
- Input: s="leetcode", dict=["leet","code"] => Output: true
- Input: s="catsandog", dict=["cats","dog","sand","and","cat"] => Output: false
- Input: s="cars", dict=["car","ca","rs"] => Output: true
Edge cases: Single character, overlapping words.

Test design coverage:
- Boundary tests: smallest sizes, zeros/empty when allowed, single-row/column when applicable.
- Equivalence classes: reachable vs unreachable states, varied distributions of values.
- Pitfall regression: known wrong iteration orders or invalid transitions.
- Randomized small tests with oracle.
*/

function solve(s, dict) {
  // TODO: implement
  return false;
}


function refSolve(s, dict) {
  const set = new Set(dict);
  const n = s.length;
  const dp = Array(n + 1).fill(false);
  dp[0] = true;
  for (let i = 1; i <= n; i += 1) {
    for (let j = 0; j < i; j += 1) {
      if (dp[j] && set.has(s.slice(j, i))) {
        dp[i] = true;
        break;
      }
    }
  }
  return dp[n];
}


function assertEqual(actual, expected, message) {
  if (actual !== expected) {
    throw new Error(`Fail: ${message}. Expected ${expected}, got ${actual}`);
  }
}

function assertDeepEqual(actual, expected, message) {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) {
    throw new Error(`Fail: ${message}. Expected ${b}, got ${a}`);
  }
}

let seed = 123456789;
function rand() {
  seed = (seed * 1664525 + 1013904223) >>> 0;
  return seed;
}

function randInt(min, max) {
  return min + (rand() % (max - min + 1));
}

let passed = 0;
let failed = 0;

function run(name, fn) {
  try {
    fn();
    passed += 1;
  } catch (err) {
    failed += 1;
    console.error(`Test failed: ${name}`);
    console.error(err.message);
  }
}

run('deterministic', () => {

  assertEqual(solve("leetcode", ["leet","code"]), true, 'deterministic');
  assertEqual(solve("applepenapple", ["apple","pen"]), true, 'deterministic');
  assertEqual(solve("catsandog", ["cats","dog","sand","and","cat"]), false, 'deterministic');
  assertEqual(solve("cars", ["car","ca","rs"]), true, 'deterministic');
  assertEqual(solve("aaaaaaa", ["aaaa","aaa"]), true, 'deterministic');
  assertEqual(solve("a", ["b"]), false, 'deterministic');

});

run('random', () => {

  const letters = ['a', 'b', 'c'];
for (let t = 0; t < 20; t += 1) {
  const wordCount = randInt(1, 6);
  const dict = Array.from({ length: wordCount }, () => {
    const len = randInt(1, 4);
    return Array.from({ length: len }, () => letters[randInt(0, 2)]).join('');
  });
  const sLen = randInt(1, 8);
  const s = Array.from({ length: sLen }, () => letters[randInt(0, 2)]).join('');
  const expected = refSolve(s, dict);
  assertEqual(solve(s, dict), expected, 'random word break');
}

});

console.log(`Passed: ${passed}, Failed: ${failed}`);

if (failed > 0) process.exit(1);

module.exports = { solve };

